### ISSUE TITLE
Performance bottlenecks

### ISSUE THREAD

**Issue by eboileau** [dieterich-lab/scimodom#55]
### A clear and concise description of what the bug is.

We're running into poor database performance with as little as a few dataset. This can already be seen on [scimodom-beta](https://scimodom-beta.dieterichlab.org/) when sorting the table. This became apparent locally when modifying the query to join genomic annotations, check also [Search: selection query does not work](https://github.com/dieterich-lab/scimodom/issues/53). 

### What browser were you using?

Firefox

### What version of Sci-ModoM were you using?

dev

**Comment by eboileau** [comment]
Handling of annotation has to be redesigning together with search queries. Consider [Many to Many](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#many-to-many), or [Association Object](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#association-object).

**Comment by eboileau** [comment]
"Split" `Data`, `GenomicAnnotation` (and bridge table) into as many models as species. But this means we need to update the schema before adding new species to the DB.

**Comment by eboileau** [comment]
New data model, added indexes, redesigned queries. Performance is better, but not as good as expected.

**Comment by eboileau** [comment]
The Search is now more restrictive: one modification > one organism (incl. cell/tissue) > one or more technologies > coordinate filters. This can speed up the query quite a bit. 

However for selected combinations this takes over ~20s e.g. mRNA/m6A > Animalia/H.Sapiens/HeLa > tick all available technologies. 

The development DB is small, and I guess there is probably a large number of dataset that could satisfy these query criteria: m6A + HeLa, for any number of technologies. This would make the number of possible records much bigger than it is now, even for just one technology. Difficult to estimate, but then in the worst case, query time will certainly go above 30s.

---

[ORM models](https://github.com/dieterich-lab/scimodom/blob/dev/server/src/scimodom/database/models.py) contain indexes on `JOIN`, `WHERE` and `SORT_BY`.

An index is also added on all FKs (some engines may do it automatically, I prefer to define it explicitly).

For the annotation table a multi index is created
https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/database/models.py#L219
where the order is preserved in the query https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/api/public.py#L195 (we make sure that all biotypes are included by default if none is selected).

For Data, a multi index is created https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/database/models.py#L333 This helps for default sort (chrom.asc, start.asc) https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/api/public.py#L212 but it becomes difficult to address all possible sorting combinations, as users may choose to sort by chrom, start, score, coverage, and frequency. There is at least individual indexes on score, coverage, and frequency.  We do not create functional Indexes e.g. ascending or descending. I'm not sure how we can do any better here and remain simple.

The DataAnnotation table has a double index https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/database/models.py#L372 which is used in the big initial `JOIN` https://github.com/dieterich-lab/scimodom/blob/3ba3ec0a2a751d1caa6499a26c629d6a2dd71882/server/src/scimodom/api/public.py#L157.

**Comment by CDieterich** [comment]
Hi @eboileau - can you please give some more details on how much time is spent on what part of the query..
afaik.. the query is a plain dump of both data sets.. 
so 20 seconds are way too long.
In other words what is the benefit of having the Search Function over a simple Download ?
Btw, the export function just exports what is visible not the entire table.

**Comment by CDieterich** [comment]
Another observation.. the filter functions for genes, biotype, etc. behave strangely.. I think, I need to show you this in person. Where can I find the database schema ?

**Comment by HaraldWilhelmi** [comment]
Agreed with Etienne that I will look closer at the following example:

* Select 'Search'
* RNA mod.: m6a
* Organism: HeLa (Animalia > Home Sapiens > HeLA)
* Technology: Select all

**Comment by HaraldWilhelmi** [comment]
API query: https://scimodom-beta.dieterichlab.org/api/v0/search?selection=1&selection=4&taxid=9606&chromStart=0&firstRecord=0&maxRecords=10

That took almost 25 s.

**Comment by HaraldWilhelmi** [comment]
For some strange reason I did not got any results until I reduced the threshold to 3 seconds:

```
MariaDB [(none)]> SET GLOBAL long_query_time=3.0;
```

Now I see two queries, which sum up the time in question:


```
# Time: 240312 13:35:33
# User@Host: scimodom[scimodom] @  [10.89.0.7]
# Thread_id: 24  Schema: scimodom  QC_hit: No
# Query_time: 11.063152  Lock_time: 0.000119  Rows_sent: 1  Rows_examined: 12254730
# Rows_affected: 0  Bytes_sent: 70
use scimodom;
SET timestamp=1710250533;
SELECT count(*) AS count_1 
FROM (SELECT data.id AS id 
FROM data_annotation INNER JOIN data ON data.id = data_annotation.data_id INNER JOIN genomic_annotation ON genomic_annotation.id = data_annotation.gene_id INNER JOIN association ON association.id = data.association_id INNER JOIN selection ON selection.id = association.selection_id INNER JOIN technology ON technology.id = selection.technology_id 
WHERE association.selection_id IN (1, 4) AND genomic_annotation.annotation_id = 1 AND genomic_annotation.biotype IN ('IG_C_gene', 'IG_D_gene', 'IG_J_gene', 'IG_J_pseudogene', 'IG_V_gene', 'IG_V_pseudogene', 'IG_gene', 'IG_pseudogene', 'LRG_gene', 'Mt_rRNA', 'Mt_tRNA', 'Mt_tRNA_pseudogene', 'RNA-Seq_gene', 'TEC', 'TR_gene', 'TR_pseudogene', 'ambiguous_orf', 'cdna_update', 'cDNA', 'disrupted_domain', 'EST', 'lincRNA', 'miRNA', 'miRNA_pseudogene', 'misc_RNA', 'misc_RNA_pseudogene', 'ncRNA', 'non_coding', 'nonsense_mediated_decay', 'polymorphic_pseudogene', 'processed_pseudogene', 'processed_transcript', 'protein_coding', 'pseudogene', 'rRNA', 'rRNA_pseudogene', 'retained_intron', 'scRNA_pseudogene', 'snRNA', 'snRNA_pseudogene', 'snlRNA', 'snoRNA', 'snoRNA_pseudogene', 'tRNA', 'tRNA_pseudogene', 'transcribed_processed_pseudogene', 'transcribed_unprocessed_pseudogene', 'unitary_pseudogene', 'unprocessed_pseudogene', 'ncRNA_host', 'TR_V_pseudogene', 'TR_V_gene', 'IG_C_pseudogene', 'TR_C_gene', 'TR_J_gene', 'protein_coding_in_progress', 'IG_M_gene', 'IG_Z_gene', '3prime_overlapping_ncRNA', 'antisense_RNA', 'scRNA', 'RNase_MRP_RNA', 'RNase_P_RNA', 'telomerase_RNA', 'sense_intronic', 'sense_overlapping', 'TR_D_gene', 'TR_J_pseudogene', 'ncbi_pseudogene', 'non_stop_decay', 'pre_miRNA', 'tmRNA', 'SRP_RNA', 'ribozyme', 'ncRNA_pseudogene', 'IG_LV_gene', 'translated_processed_pseudogene', 'nontranslating_CDS', 'translated_unprocessed_pseudogene', 'mRNA', 'artifact', 'class_I_RNA', 'class_II_RNA', 'known_ncRNA', 'transcribed_unitary_pseudogene', 'piRNA', 'scaRNA', 'sRNA', 'antitoxin', 'macro_lncRNA', 'IG_D_pseudogene', 'guide_RNA', 'Y_RNA', 'transposable_element', 'bidirectional_promoter_lncRNA', 'unknown_likely_coding', 'lncRNA', 'aligned_transcript', 'antisense', 'vault_RNA', 'rnaseq_putative_cds', 'transcribed_pseudogene', 'other', 'pseudogene_with_CDS', 'misc_non_coding', 'miRNA_primary_transcript', 'protein_coding_CDS_not_defined', 'protein_coding_LoF') GROUP BY data_annotation.data_id) AS anon_1;
# Time: 240312 13:35:44
# User@Host: scimodom[scimodom] @  [10.89.0.7]
# Thread_id: 24  Schema: scimodom  QC_hit: No
# Query_time: 13.693916  Lock_time: 0.000169  Rows_sent: 10  Rows_examined: 15477348
# Rows_affected: 0  Bytes_sent: 1780
SET timestamp=1710250544;
SELECT data.id, data.chrom, data.start, data.end, data.name, data.score, data.strand, data.coverage, data.frequency, group_concat(DISTINCT data_annotation.feature) AS feature, group_concat(DISTINCT genomic_annotation.biotype) AS gene_biotype, group_concat(DISTINCT genomic_annotation.name) AS gene_name, technology.tech, association.dataset_id 
FROM data_annotation INNER JOIN data ON data.id = data_annotation.data_id INNER JOIN genomic_annotation ON genomic_annotation.id = data_annotation.gene_id INNER JOIN association ON association.id = data.association_id INNER JOIN selection ON selection.id = association.selection_id INNER JOIN technology ON technology.id = selection.technology_id 
WHERE association.selection_id IN (1, 4) AND genomic_annotation.annotation_id = 1 AND genomic_annotation.biotype IN ('IG_C_gene', 'IG_D_gene', 'IG_J_gene', 'IG_J_pseudogene', 'IG_V_gene', 'IG_V_pseudogene', 'IG_gene', 'IG_pseudogene', 'LRG_gene', 'Mt_rRNA', 'Mt_tRNA', 'Mt_tRNA_pseudogene', 'RNA-Seq_gene', 'TEC', 'TR_gene', 'TR_pseudogene', 'ambiguous_orf', 'cdna_update', 'cDNA', 'disrupted_domain', 'EST', 'lincRNA', 'miRNA', 'miRNA_pseudogene', 'misc_RNA', 'misc_RNA_pseudogene', 'ncRNA', 'non_coding', 'nonsense_mediated_decay', 'polymorphic_pseudogene', 'processed_pseudogene', 'processed_transcript', 'protein_coding', 'pseudogene', 'rRNA', 'rRNA_pseudogene', 'retained_intron', 'scRNA_pseudogene', 'snRNA', 'snRNA_pseudogene', 'snlRNA', 'snoRNA', 'snoRNA_pseudogene', 'tRNA', 'tRNA_pseudogene', 'transcribed_processed_pseudogene', 'transcribed_unprocessed_pseudogene', 'unitary_pseudogene', 'unprocessed_pseudogene', 'ncRNA_host', 'TR_V_pseudogene', 'TR_V_gene', 'IG_C_pseudogene', 'TR_C_gene', 'TR_J_gene', 'protein_coding_in_progress', 'IG_M_gene', 'IG_Z_gene', '3prime_overlapping_ncRNA', 'antisense_RNA', 'scRNA', 'RNase_MRP_RNA', 'RNase_P_RNA', 'telomerase_RNA', 'sense_intronic', 'sense_overlapping', 'TR_D_gene', 'TR_J_pseudogene', 'ncbi_pseudogene', 'non_stop_decay', 'pre_miRNA', 'tmRNA', 'SRP_RNA', 'ribozyme', 'ncRNA_pseudogene', 'IG_LV_gene', 'translated_processed_pseudogene', 'nontranslating_CDS', 'translated_unprocessed_pseudogene', 'mRNA', 'artifact', 'class_I_RNA', 'class_II_RNA', 'known_ncRNA', 'transcribed_unitary_pseudogene', 'piRNA', 'scaRNA', 'sRNA', 'antitoxin', 'macro_lncRNA', 'IG_D_pseudogene', 'guide_RNA', 'Y_RNA', 'transposable_element', 'bidirectional_promoter_lncRNA', 'unknown_likely_coding', 'lncRNA', 'aligned_transcript', 'antisense', 'vault_RNA', 'rnaseq_putative_cds', 'transcribed_pseudogene', 'other', 'pseudogene_with_CDS', 'misc_non_coding', 'miRNA_primary_transcript', 'protein_coding_CDS_not_defined', 'protein_coding_LoF') GROUP BY data_annotation.data_id ORDER BY data.chrom ASC, data.start ASC 
 LIMIT 0, 10;
```

**Comment by HaraldWilhelmi** [comment]
Formatting the first statement nicer:

```
SELECT count(*) AS count_1
FROM (
    SELECT data.id AS id 
    FROM data_annotation
    INNER JOIN data
        ON data.id = data_annotation.data_id
    INNER JOIN genomic_annotation
        ON genomic_annotation.id = data_annotation.gene_id
    INNER JOIN association
        ON association.id = data.association_id
    INNER JOIN selection
        ON selection.id = association.selection_id
    INNER JOIN technology
    ON technology.id = selection.technology_id
    WHERE association.selection_id IN (1, 4)
        AND genomic_annotation.annotation_id = 1
        AND genomic_annotation.biotype IN (
            'IG_C_gene', 'IG_D_gene', 'IG_J_gene', 'IG_J_pseudogene', 'IG_V_gene', 'IG_V_pseudogene', 'IG_gene', 'IG_pseudogene',
            'LRG_gene', 'Mt_rRNA', 'Mt_tRNA', 'Mt_tRNA_pseudogene', 'RNA-Seq_gene', 'TEC', 'TR_gene', 'TR_pseudogene', 'ambiguous_orf',
            'cdna_update', 'cDNA', 'disrupted_domain', 'EST', 'lincRNA', 'miRNA', 'miRNA_pseudogene', 'misc_RNA', 'misc_RNA_pseudogene',
            'ncRNA', 'non_coding', 'nonsense_mediated_decay', 'polymorphic_pseudogene', 'processed_pseudogene', 'processed_transcript',
            'protein_coding', 'pseudogene', 'rRNA', 'rRNA_pseudogene', 'retained_intron', 'scRNA_pseudogene', 'snRNA', 'snRNA_pseudogene',
            'snlRNA', 'snoRNA', 'snoRNA_pseudogene', 'tRNA', 'tRNA_pseudogene', 'transcribed_processed_pseudogene',
            'transcribed_unprocessed_pseudogene', 'unitary_pseudogene', 'unprocessed_pseudogene', 'ncRNA_host', 'TR_V_pseudogene',
            'TR_V_gene', 'IG_C_pseudogene', 'TR_C_gene', 'TR_J_gene', 'protein_coding_in_progress', 'IG_M_gene', 'IG_Z_gene',
            '3prime_overlapping_ncRNA', 'antisense_RNA', 'scRNA', 'RNase_MRP_RNA', 'RNase_P_RNA', 'telomerase_RNA', 'sense_intronic',
            'sense_overlapping', 'TR_D_gene', 'TR_J_pseudogene', 'ncbi_pseudogene', 'non_stop_decay', 'pre_miRNA', 'tmRNA', 'SRP_RNA',
            'ribozyme', 'ncRNA_pseudogene', 'IG_LV_gene', 'translated_processed_pseudogene', 'nontranslating_CDS',
            'translated_unprocessed_pseudogene', 'mRNA', 'artifact', 'class_I_RNA', 'class_II_RNA', 'known_ncRNA',
            'transcribed_unitary_pseudogene', 'piRNA', 'scaRNA', 'sRNA', 'antitoxin', 'macro_lncRNA', 'IG_D_pseudogene', 'guide_RNA',
            'Y_RNA', 'transposable_element', 'bidirectional_promoter_lncRNA', 'unknown_likely_coding', 'lncRNA', 'aligned_transcript',
            'antisense', 'vault_RNA', 'rnaseq_putative_cds', 'transcribed_pseudogene', 'other', 'pseudogene_with_CDS', 'misc_non_coding',
            'miRNA_primary_transcript', 'protein_coding_CDS_not_defined', 'protein_coding_LoF')
    GROUP BY data_annotation.data_id
) AS anon_1
```

Select returns about 1.4 M (1462239). The result selected about 62700 rows from genomic_annotation.

**Comment by HaraldWilhelmi** [comment]
Just to be on the save side I checked again the indexes:
```
MariaDB [scimodom]> SHOW INDEXES FROM genomic_annotation;
+--------------------+------------+-------------------------------------+--------------+---------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| Table              | Non_unique | Key_name                            | Seq_in_index | Column_name   | Collation | Cardinality | Sub_part | Packed | Null | Index_type | Comment | Index_comment | Ignored |
+--------------------+------------+-------------------------------------+--------------+---------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
| genomic_annotation |          0 | PRIMARY                             |            1 | id            | A         |      118925 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| genomic_annotation |          1 | idx_genomic                         |            1 | annotation_id | A         |           1 |     NULL | NULL   |      | BTREE      |         |               | NO      |
| genomic_annotation |          1 | idx_genomic                         |            2 | biotype       | A         |          88 |     NULL | NULL   | YES  | BTREE      |         |               | NO      |
| genomic_annotation |          1 | idx_genomic                         |            3 | name          | A         |      118925 |     NULL | NULL   | YES  | BTREE      |         |               | NO      |
| genomic_annotation |          1 | ix_genomic_annotation_annotation_id |            1 | annotation_id | A         |           1 |     NULL | NULL   |      | BTREE      |         |               | NO      |
+--------------------+------------+-------------------------------------+--------------+---------------+-----------+-------------+----------+--------+------+------------+---------+---------------+---------+
```

Don't see anything we could add there.

**Comment by eboileau** [comment]
We decided to move the table filters (Feature, Biotype, Gene) as top filters with sensible defaults. Live filtering of the
table is as such removed. In general, this will have no significant effect on the query time, unless a user intentionally restricts the search *e.g.* to certain features or a given gene.

Features and biotypes (do they change with organism?) are known. For gene, if we want to have auto completion or something similar, we need to think how to "optimize" this, as the gene set will change with every query, and unless we
handle that in a smart way, this will significantly impact performance again. But to start, we can just implement a regular input.

We could eventually warn users of potentially large queries.

Closing this for now, and moving that to Q2 tasks.
